src/base/bittorrent/dbresumedatastorage.cpp

```diff
@@ -0,0 +1,573 @@
+#include "base/utils/fs.h"
+#include "base/utils/string.h"
+#include "infohash.h"
+#include "loadtorrentparams.h"
+#include "torrentinfo.h"
+
+namespace
+{
+    const char DB_CONNECTION_NAME[] = "ResumeDataStorage";
+
+    const int DB_VERSION = 1;
+
+    const char DB_TABLE_META[] = "meta";
+    const char DB_TABLE_TORRENTS[] = "torrents";
+
+    struct Column
+    {
+        QString name;
+        QString placeholder;
+    };
+
+    Column makeColumn(const QString &columnName)
+    {
+        return {columnName, (QLatin1Char(':') + columnName)};
+    }
+
+    const Column DB_COLUMN_ID = makeColumn(QLatin1String("id"));
+    const Column DB_COLUMN_TORRENT_ID = makeColumn(QLatin1String("torrent_id"));
+    const Column DB_COLUMN_QUEUE_POSITION = makeColumn(QLatin1String("queue_position"));
+    const Column DB_COLUMN_NAME = makeColumn(QLatin1String("name"));
+    const Column DB_COLUMN_CATEGORY = makeColumn(QLatin1String("category"));
+    const Column DB_COLUMN_TAGS = makeColumn(QLatin1String("tags"));
+    const Column DB_COLUMN_TARGET_SAVE_PATH = makeColumn(QLatin1String("target_save_path"));
+    const Column DB_COLUMN_CONTENT_LAYOUT = makeColumn(QLatin1String("content_layout"));
+    const Column DB_COLUMN_RATIO_LIMIT = makeColumn(QLatin1String("ratio_limit"));
+    const Column DB_COLUMN_SEEDING_TIME_LIMIT = makeColumn(QLatin1String("seeding_time_limit"));
+    const Column DB_COLUMN_HAS_OUTER_PIECES_PRIORITY = makeColumn(QLatin1String("has_outer_pieces_priority"));
+    const Column DB_COLUMN_HAS_SEED_STATUS = makeColumn(QLatin1String("has_seed_status"));
+    const Column DB_COLUMN_OPERATING_MODE = makeColumn(QLatin1String("operating_mode"));
+    const Column DB_COLUMN_STOPPED = makeColumn(QLatin1String("stopped"));
+    const Column DB_COLUMN_RESUMEDATA = makeColumn(QLatin1String("libtorrent_resume_data"));
+    const Column DB_COLUMN_METADATA = makeColumn(QLatin1String("metadata"));
+    const Column DB_COLUMN_VALUE = makeColumn(QLatin1String("value"));
+
+    template <typename LTStr>
+    QString fromLTString(const LTStr &str)
+    {
+        return QString::fromUtf8(str.data(), static_cast<int>(str.size()));
+    }
+
+    QString quoted(const QString &name)
+    {
+        const QLatin1Char QUOTE {'`'};
+
+        return QUOTE + name + QUOTE;
+    }
+
+    QString makeCreateTableStatement(const QString &tableName, const QStringList &items)
+    {
+        return QString::fromLatin1("CREATE TABLE %1 (%2)").arg(quoted(tableName), items.join(", "));
+    }
+
+    QString makeInsertStatement(const QString &tableName, const QVector<Column> &columns)
+    {
+        QStringList names;
+        names.reserve(columns.size());
+        QStringList values;
+        values.reserve(columns.size());
+        for (const Column &column : columns)
+        {
+            names.append(quoted(column.name));
+            values.append(column.placeholder);
+        }
+
+        const QString jointNames = names.join(QLatin1String(", "));
```

Comment:
would it be possible to use `names.join(QLatin1Char(','));`?

Reply:
This, of course, is possible, if we are not talking about generating user-friendly/readable SQL scripts.

Reply:
But what is the significant benefit of this, other than saving a couple of bytes in the resulting string?

Reply:
>This, of course, is possible, if we are not talking about generating user-friendly/readable SQL scripts.

It depends on what our goals here, are the SQL commands intended to be inspected by human under normal operation? If not, I would not take command readability into consideration.